Plutarch/Builtin.hs

```diff
@@ -225,8 +227,18 @@ instance PConstant (PAsDataLifted a) where
 
 instance PUnsafeLiftDecl (PAsData a) where type PLifted (PAsData a) = PAsDataLifted a
 
-pforgetData :: Term s (PAsData a) -> Term s PData
-pforgetData = punsafeCoerce
+newtype Helper1 (a :: PType) (s :: S) = Helper1 (a s)
+
+pforgetData :: forall s a. Term s (PAsData a) -> Term s PData
+pforgetData x = coerce $ pforgetData' Proxy (coerce x :: Term s (Helper1 (PAsData a)))
+
+-- | Like 'pforgetData', except it works for complex types.
+pforgetData' :: forall a (p :: PType -> PType) s. Proxy p -> Term s (p (PAsData a)) -> Term s (p PData)
+pforgetData' Proxy = punsafeCoerce
+
+-- | Like 'pforgetData', except it works for complex types.
+prememberData :: forall (p :: PType -> PType) s. Proxy p -> Term s (p PData) -> Term s (p (PAsData PData))
```

Comment:
these two functions are very nice (though this one's haddock is wrong), but it'd be nice to have this for all  `PIsDataRepr a` as well.

```hs
PIsDataRepr => Term s (p (PAsData a)) -> Term s (p a)
```
and such, perhaps not as a separate function though.